src/compiler/checker.ts

```diff
@@ -4415,6 +4415,21 @@ namespace ts {
 
                 return Debug.fail("Should be unreachable.");
 
+
+                function typeToTypeNodeOrCircularityEllision(type: Type) {
+                    if (type.flags & TypeFlags.Union) {

```

Comment:
Unions can't normally be circularly created - excepting when they appear as the result of a conditional, and that conditional self references as a result of an indexed access on an object (the conditional will defer the evaluation). The union in question here ends up being the result of the indexing operation. We can't do all types because anonymous objects already do a very specialized form of circular reference checking using this cache, so including them would end up checking/inserting the cache twice and cause bogus `any` printbacks. Unions, however, do not use this cache anywhere yet~